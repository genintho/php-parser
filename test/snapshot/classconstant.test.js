const parser = require("../main");

describe("classconstant", () => {
  it("simple", () => {
    expect(
      parser.parseEval('class Foo { const CONSTANT = "Hello world!"; }'),
    ).toMatchSnapshot();
  });
  it("multiple", () => {
    expect(
      parser.parseEval(
        'class Foo { const CONSTANT = "Hello world!", OTHER_CONSTANT = "Other hello world!"; }',
      ),
    ).toMatchSnapshot();
  });
  it("public", () => {
    expect(
      parser.parseEval('class Foo { public const CONSTANT = "Hello world!"; }'),
    ).toMatchSnapshot();
  });
  it("protected", () => {
    expect(
      parser.parseEval(
        'class Foo { protected const CONSTANT = "Hello world!"; }',
      ),
    ).toMatchSnapshot();
  });
  it("private", () => {
    expect(
      parser.parseEval(
        'class Foo { private const CONSTANT = "Hello world!"; }',
      ),
    ).toMatchSnapshot();
  });
  it("final", () => {
    expect(
      parser.parseEval(
        'class Foo { final public const CONSTANT = "Hello world!"; }',
      ),
    ).toMatchSnapshot();
  });
  it("type hinted (supported)", () => {
    expect(
      parser.parseEval(
        `class Foo {
              public const CON_1 = "Hello world!";
              const CON_2 = "Hello world!";
              const string CON_3 = "Hello world!";
              public const string CON_4 = "Hello world!";
              public const string|int CON_5 = "Hello world!";
              const string|int CON_6 = "Hello world!";
              }`,
        { parser: { version: 803 } },
      ),
    ).toMatchSnapshot();
  });

  it("type hinted list of constant with mixed type", () => {
    expect(
      parser.parseEval(
        `class Foo {
                  public const string PUB1 = 'bar', PUB2 = 'bar2', int PUB3 = 1;
                  private const PRI1 = 'baz', string PRI2 = 'baz2';
                  private const string BOB = 'baz2';
                }`,
        { parser: { version: 803 } },
      ),
    ).toMatchSnapshot();
  });

  it("type hinted (unsupported)", () => {
    expect(() =>
      parser.parseEval(
        'class Foo { public const  DDCONSTANT = "Hello world!"; public const string CONSTANT = "Hello world!"; }',
        { parser: { version: 802 } },
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it("accept the constant name 'list'", () => {
    expect(
      parser.parseEval(
        `class Foo {
              const list = "Hello world!";
              }`,
        { parser: { version: 803 } },
      ),
    ).toMatchSnapshot();
  });
});
