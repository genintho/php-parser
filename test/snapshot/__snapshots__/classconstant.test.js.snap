// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classconstant accept the constant name 'list' 1`] = `
Program {
  "children": [
    Class {
      "attrGroups": [],
      "body": [
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "list",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "isReadonly": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Foo",
      },
    },
  ],
  "errors": [],
  "kind": "program",
}
`;

exports[`classconstant final 1`] = `
Program {
  "children": [
    Class {
      "attrGroups": [],
      "body": [
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "CONSTANT",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": true,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "public",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "isReadonly": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Foo",
      },
    },
  ],
  "errors": [],
  "kind": "program",
}
`;

exports[`classconstant multiple 1`] = `
Program {
  "children": [
    Class {
      "attrGroups": [],
      "body": [
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "CONSTANT",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "OTHER_CONSTANT",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Other hello world!"",
                "unicode": false,
                "value": "Other hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "isReadonly": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Foo",
      },
    },
  ],
  "errors": [],
  "kind": "program",
}
`;

exports[`classconstant private 1`] = `
Program {
  "children": [
    Class {
      "attrGroups": [],
      "body": [
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "CONSTANT",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "private",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "isReadonly": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Foo",
      },
    },
  ],
  "errors": [],
  "kind": "program",
}
`;

exports[`classconstant protected 1`] = `
Program {
  "children": [
    Class {
      "attrGroups": [],
      "body": [
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "CONSTANT",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "protected",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "isReadonly": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Foo",
      },
    },
  ],
  "errors": [],
  "kind": "program",
}
`;

exports[`classconstant public 1`] = `
Program {
  "children": [
    Class {
      "attrGroups": [],
      "body": [
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "CONSTANT",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "public",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "isReadonly": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Foo",
      },
    },
  ],
  "errors": [],
  "kind": "program",
}
`;

exports[`classconstant simple 1`] = `
Program {
  "children": [
    Class {
      "attrGroups": [],
      "body": [
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "CONSTANT",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "isReadonly": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Foo",
      },
    },
  ],
  "errors": [],
  "kind": "program",
}
`;

exports[`classconstant type hinted (supported) 1`] = `
Program {
  "children": [
    Class {
      "attrGroups": [],
      "body": [
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "CON_1",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "public",
        },
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "name": Identifier {
                "kind": "identifier",
                "name": "CON_2",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "",
        },
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "loc": TypeReference {
                "kind": "typereference",
                "name": "string",
                "raw": "string",
              },
              "name": Identifier {
                "kind": "identifier",
                "name": "CON_3",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "",
        },
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "loc": TypeReference {
                "kind": "typereference",
                "name": "string",
                "raw": "string",
              },
              "name": Identifier {
                "kind": "identifier",
                "name": "CON_4",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "public",
        },
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "loc": UnionType {
                "kind": "uniontype",
                "name": null,
                "types": [
                  TypeReference {
                    "kind": "typereference",
                    "name": "string",
                    "raw": "string",
                  },
                  TypeReference {
                    "kind": "typereference",
                    "name": "int",
                    "raw": "int",
                  },
                ],
              },
              "name": Identifier {
                "kind": "identifier",
                "name": "CON_5",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "public",
        },
        ClassConstant {
          "attrGroups": undefined,
          "constants": [
            Constant {
              "kind": "constant",
              "loc": UnionType {
                "kind": "uniontype",
                "name": null,
                "types": [
                  TypeReference {
                    "kind": "typereference",
                    "name": "string",
                    "raw": "string",
                  },
                  TypeReference {
                    "kind": "typereference",
                    "name": "int",
                    "raw": "int",
                  },
                ],
              },
              "name": Identifier {
                "kind": "identifier",
                "name": "CON_6",
              },
              "value": String {
                "isDoubleQuote": true,
                "kind": "string",
                "raw": ""Hello world!"",
                "unicode": false,
                "value": "Hello world!",
              },
            },
          ],
          "final": false,
          "kind": "classconstant",
          "nullable": [],
          "type": null,
          "visibility": "",
        },
      ],
      "extends": null,
      "implements": null,
      "isAbstract": false,
      "isAnonymous": false,
      "isFinal": false,
      "isReadonly": false,
      "kind": "class",
      "name": Identifier {
        "kind": "identifier",
        "name": "Foo",
      },
    },
  ],
  "errors": [],
  "kind": "program",
}
`;

exports[`classconstant type hinted (unsupported) 1`] = `"Parse Error: Typed Class Constants requires PHP 8.3+ on line 1"`;
